File: 프로그래머스/Lv2/전력망을 둘로 나누기(X).py
# 다른 사람 풀이
from collections import deque

def solution(n, wires):
    answer = 0
    graph = [[] for _ in range(n+1)]
    
    for a,b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    def bfs(start):
        visited = [0] * (n+1)
        q = deque([start])
        visited[start] = 1
        cnt = 1
        
        while q:
            s = q.popleft()
            for i in graph[s]:
                if not visited[i]:
                    q.append(i)
                    visited[i] = 1
                    cnt += 1
        return cnt
            
    answer = n
    
    for a,b in wires:
        #graph에서 remove
        graph[a].remove(b)
        graph[b].remove(a)
        
        answer = min(abs(bfs(a) - bfs(b)), answer)
        
        #다시 append
        graph[a].append(b)
        graph[b].append(a)
    
    return answer

# 다른 사람 풀이 2
def solution2(n, wires):
    ans = n
    for sub in (wires[i+1:] + wires[:i] for i in range(len(wires))):
        s = set(sub[0])
        [s.update(v) for _ in sub for v in sub if set(v) & s] # 집합연산자 & : 교집합 연산,   집합연산자 update : 여러데이터를 한번에 추가
        ans = min(ans, abs(2 * len(s) - n))
    return ans

print(solution(9, [[1,3],[2,3],[3,4],[4,5],[4,6],[4,7],[7,8],[7,9]]))
print(solution(4, [[1,2],[2,3],[3,4]]))
print(solution2(7, [[1,2],[2,7],[3,7],[3,4],[4,5],[6,7]]))

File: 프로그래머스/Lv2/전력망을 둘로 나누기(X).js
// 다른 사람 풀이 1
function solution(n, wires) {
    var answer = n;
    
    for(let i=0; i<wires.length; i++){
        const parent = Array.from({length:n+1}, (_, i) => i);
        for(let j=0; j<wires.length; j++){
            if(i == j) continue;
            union_parent(parent, wires[j][0], wires[j][1]);
        }
        
        for(let wire of wires){
            let a = wire[0];
            let b = wire[1];
            
            parent[a] = find_parent(parent, a);
            parent[b] = find_parent(parent, b); 
        }
        
        s1 = parent.filter((el) => el === parent[wires[i][0]]).length;
        s2 = parent.filter((el) => el === parent[wires[i][1]]).length;
        answer = Math.min(answer, Math.abs(s1- s2));
    }
    
    return answer;
}

function union_parent(parent, a, b){
    a = find_parent(parent, a);
    b = find_parent(parent, b);
    
    if(a < b) parent[b] = a;
    else parent[a] = b;
}

function find_parent(parent, x){
    if(parent[x] !== x){
        return find_parent(parent, parent[x]);
    }
    
    return parent[x];
}

// 다른 사람 풀이 2
function solution2(n, wires) {
    let answer = n;
    const conn = Array.from({length: n+1}, () => []);
    
    for(let wire of wires){
        conn[wire[0]].push(wire[1]);
        conn[wire[1]].push(wire[0]);
    }
    
    for(let wire of wires){
        conn[wire[0]] = conn[wire[0]].filter(el => el !== wire[1]);
        conn[wire[1]] = conn[wire[1]].filter(el => el !== wire[0]);
            
        answer = Math.min(answer, Math.abs(bfs(wire[0], n, conn) - bfs(wire[1], n, conn)));
        
        conn[wire[0]].push(wire[1]);
        conn[wire[1]].push(wire[0]);
    }
    
    return answer;
}

function bfs(start, n, conn){
    let cnt = 0;
    const q = [start];
    const visited = Array.from({length:n+1}, () => false);
    
    while(q.length !== 0){
        let temp = q.pop();
        
        if(visited[temp] === false){
            q.push(...conn[temp]);
            visited[temp] = true;
            cnt++;
        }
    }
    
    return cnt;
}

console.log(solution(9, [[1,3],[2,3],[3,4],[4,5],[4,6],[4,7],[7,8],[7,9]]))
console.log(solution(4, [[1,2],[2,3],[3,4]]))
console.log(solution2(7, [[1,2],[2,7],[3,7],[3,4],[4,5],[6,7]]))